src/omp/exec_dist.f90

```diff
@@ -61,5 +61,134 @@ subroutine exec_dist_tds_compact( &
 
    end subroutine exec_dist_tds_compact
 
+
+   subroutine exec_dist_transeq_compact(&
+      rhs, du, dud, d2u, &
+      du_send_s, du_send_e, du_recv_s, du_recv_e, &
+      dud_send_s, dud_send_e, dud_recv_s, dud_recv_e, &
+      d2u_send_s, d2u_send_e, d2u_recv_s, d2u_recv_e, &
+      u, u_recv_s, u_recv_e, &
+      v, v_recv_s, v_recv_e, &
+      tdsops_du, tdsops_dud, tdsops_d2u, nu, nproc, pprev, pnext, n_block)
+
+      implicit none
+
+      ! du = d(u)
+      real(dp), dimension(:, :, :), intent(out) :: rhs, du, dud, d2u
+
+      ! The ones below are intent(out) just so that we can write data in them,
+      ! not because we actually need the data they store later where this
+      ! subroutine is called. We absolutely don't care about the data they pass back
+      real(dp), dimension(:, :, :), intent(out) :: &
+         du_send_s, du_send_e, du_recv_s, du_recv_e
+      real(dp), dimension(:, :, :), intent(out) :: &
+         dud_send_s, dud_send_e, dud_recv_s, dud_recv_e
+      real(dp), dimension(:, :, :), intent(out) :: &
+         d2u_send_s, d2u_send_e, d2u_recv_s, d2u_recv_e
+
+      real(dp), dimension(:, :, :), intent(in) :: u, u_recv_s, u_recv_e
+      real(dp), dimension(:, :, :), intent(in) :: v, v_recv_s, v_recv_e
+
+      type(tdsops_t), intent(in) :: tdsops_du, tdsops_dud, tdsops_d2u
+
+      real(dp), dimension(:, :), allocatable :: ud, ud_recv_s, ud_recv_e
+      real(dp) :: nu
+      integer, intent(in) :: nproc, pprev, pnext
+      integer, intent(in) :: n_block
+
+      integer :: n_data, n_halo
+      integer :: k, i, j, n
+
+      ! TODO: don't hardcode n_halo
+      n_halo = 4
+      n = tdsops_d2u%n
+      n_data = SZ*n_block
+
+      allocate(ud(SZ, n))
+      allocate(ud_recv_e(SZ, n_halo))
+      allocate(ud_recv_s(SZ, n_halo))
+
+      !$omp parallel do
```

Comment:
We make use of `ud`, `ud_recv_e` and `ud_recv_s` arrays defined and allocated above here inside the parallel do loop. We need a minor fix here to ensure we get correct result and not result in race conditions. I suspect that the current state can produce wrong result in some cases, and due to race conditions it is probably not as fast as it can be.

When you enter a OpenMP parallel region you instantly have OMP_NUM_THREADS many processes working together. And the variables and arrays you defined outside a parallel region can be accessed by any of the threads. For example, we only have one `du` field, and all threads in this parallel region can read/write this array. But because pass a unique part of these arrays for each process (via the loop over k), there is no problem with them. But for example, if you write `du(1,1,1)  = 0`, all of the threads will do the exact same operation for you, its a very important thing to consider when using OpenMP.

The problem here as you can guess now is with the temporary arrays, (`ud` ..). Unless we specify, all threads will try to make use of the very same memory when we try to write or read from `ud`. We don't want this because these threads are working on different parts of the domain and need their own temporary storages. It is easy to fix this, just replace 
`!$omp parallel do` with
`!$omp parallel do private(ud, ud_recv_e, ud_recv_s)`
so that all threads have their own copy of these temporary arrays.

I guess the tests pass because we have the exact sin wave in all the lines in the domain, so the temporaries are exactly the same accross domain. If the initial field is not the same accross the domain I don't think the test can pass.

example: https://github.com/xcompact3d/transeq_perf/blob/9849028f940200800b449f80d394fde67729a585/vectranseq.f90#L215

Reply:
ah yes of course, well spotted. My openMP is a bit rusty and I lost the reflex to check these things. I will correct that